### ISSUE TITLE
Regent: Support discard on regions

### ISSUE THREAD

**Issue by elliottslaughter** [StanfordLegion/legion#721]
I'm creating this issue to collect discussion for the design and implementation of `discard` in Regent. This has come up a number of times in other issues since the lack of `discard` results in warnings when accessing a region for the first time.

As of d237068fb31fd711085140205850fc35f8ce9230 there is a limited fix in place, but it only applies when a task is not `inner` (otherwise no inline mappings are created).

One proposal is for `discard` to be a pseudo-privilege:

```
task f(r : region(int))
where reads writes discards(r) do
end
```

This would permit task `f` to read and write the contents of `r`, but would assert that `f` must at least overwrite each value once before reading (otherwise there will be garbage).

I know people had some other ideas, please discuss them here.

**Comment by elliottslaughter** [comment]
Copying from private discussion: @streichler has suggested that `discard` should be a verb, like `fill`. You'd do something like:

```
var r = region(...)
discard(r)
some_task(r)
```

That is, the contents of `r` will be explicitly discarded before running `some_task`.

I don't really see a way to implement this without runtime support, but it's worth keeping in mind as one of the available options.

**Comment by lightsighter** [comment]
It's worth noting that doing it like that will double the runtime overhead of discard as now there would be a whole other operation that has to go through both logical and physical analysis and everything that is associated with it. I'm very much not in favor of that.

**Comment by lightsighter** [comment]
And actually, we already have an operation like that: its called `fill`. If you want to do discard as a separate operation you can just call `fill`. :)

**Comment by streichler** [comment]
Yes, my assumption was that discard would look just like a fill to every part of the Legion runtime except the part that actually issues a Realm fill operation.  Think of discard as "fill with an undefined value".

**Comment by lightsighter** [comment]
Why would you ever want to call `fill` with an "undefined value"? I can support that, but you're still going to pay a cost for a second whole operation having to go through the operation pipeline and make a specialized view. You don't have to pay that tax today in the runtime where you can just attach discard to the privileges.

**Comment by streichler** [comment]
A fill (or a discard) has the nice property of invalidating existing instances right away, allowing them to be collected.  A discard further has the advantage that you don't do an unnecessary fill of the instance before the task that redefines the field later on.

There are many cases where the last use of a field on one timestep comes quite a ways before the redefinition of the field in the next timestep, and using write discard privileges results in having to pay for storage for that field at all times.

**Comment by lightsighter** [comment]
The first case you describe above can be done with a `fill` today. The second one you describe could be handled with `fill` with a `none` value. Nothing though can achieve the efficiency of fusing the discard with a region privilege in terms of runtime overhead, so it's never going to go away.

**Comment by opensdh** [comment]
> Can you articulate why you think it needs to happen in the same task and not in a separate operation as was proposed earlier in this thread?

As a separate operation, there would be the [additional analysis overhead](#issuecomment-582525677); moreover, the runtime would not know to suppress the warning/error for a `READ_WRITE` mapping that overlapped the invalidated region in a _later_ task.  Accomplishing that with the tools we have today (_i.e._, one `READ_WRITE` region requirement and one "adjacent" `WRITE_DISCARD` requirement) involves keeping partitions (and associated bookkeeping information) around longer (you need _p_<sub>_n_-1</sub> and _&delta;_<sub>_n_</sub> while keeping _p_<sub>_n_</sub> around for the next task) and complicates accessing the data as the _one_ (albeit only partially initialized) array that the application wants to use.

> Note that this cost will grow with O(N^2) on the number of region requirements in every task launch.

Why would you compare every pair?  It's only the `DISCARD` ones that might overlap with such a purpose, and by requiring that they come before any that overlap with them, you can surely account for them without expense significantly beyond that already required to determine that new instances need to be created with the new size (and handle the common case of no `DISCARD` very cheaply).  Our application would of course overallocate to avoid frequent use of `DISCARD` at the usual expense of a bit of wasted memory (and any time required to move it around).

**Comment by lightsighter** [comment]
> Accomplishing that with the tools we have today (i.e., one READ_WRITE region requirement and one "adjacent" WRITE_DISCARD requirement) involves keeping partitions (and associated bookkeeping information) around longer (you need pn-1 and δn while keeping pn around for the next task) and complicates accessing the data as the one (albeit only partially initialized) array that the application wants to use.

Ok, so there is no functional reason for this, because what you are proposing adds considerable complexity to the runtime implementation, which is a dimension that I don't think you are considering. If it is functionally necessary then it is unavoidable, but if it just a performance optimization then we need to weigh the tradeoff between complexity and performance.

> Why would you compare every pair?

You have to compare every discard region requirement with every non-discard region requirement. `O(MN)` where `M` is the number of discard requirements and `N` is the number of non-discard requirements is more accurate. 

> It's only the DISCARD ones that might overlap with such a purpose, and by requiring that they come before any that overlap with them, you can surely account for them without expense significantly beyond that already required to determine that new instances need to be created with the new size (and handle the common case of no DISCARD very cheaply).

You're talking about two different stages of the execution pipeline. Nothing is cheap if it requires state that spans pipeline stages, especially not pipeline stags that can be executed on different nodes for the same operation.

I will also state my opinion here for the group to consider: I personally think enforcing any kind of ABI-like semantics on region requirements runs against Legion's philosophy. Legion shouldn't be enforcing an ABI on tasks as it's just the runtime. Compilers and applications should be able to define any ABI on region requirements that they want without any constrains on the region requirements enforced by the runtime. My opinion is that it should remain that way. That of course is a separate concern to how you define "partial discard", but I do think it will (and should) constrain any implementation.
